File: python/n409_LongestPa.py
# https://leetcode-cn.com/problems/longest-palindrome/
class Solution:
    def longestPalindrome(self, s: str) -> int:
        chr_list = [0]*52
        for i in s:
            ind = ord(i) - ord('a')
            chr_list[ind] += 1
        res = odd = 0
        for j in chr_list:
            if j % 2 == 0:
                res += j
            else:
                res = res + j - 1
                odd = 1
        return res + odd


if __name__ == "__main__":
    solu = Solution()
    s = "civilwartestingwhetherthatnaptionoranynartio"
    print(solu.longestPalindrome(s))

File: cpp/n409_LongestPa.cpp
// https://leetcode-cn.com/problems/longest-palindrome/
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    int longestPalindrome(string s) {
        vector <int> sli(58, 0); // 注意asicc码中‘Z’到‘a’中间还有六个特殊字符，所以长度要58
        int res = 0, odd = 0;
        for (auto i : s){
            int ind = i - 'A';
            sli[ind] += 1;
        }
        for (size_t j = 0; j < sli.size(); ++j){
            if (sli[j] % 2 == 0) res += sli[j];
            else{
                odd = 1;
                res = res + sli[j] - 1;
            }
        }
        return (res + odd);
    }
};

int main(){
    Solution solu;
    string s = "AAAaaacderDER";
    cout << solu.longestPalindrome(s) << endl;
}